src/handler.rs

```diff
@@ -577,9 +609,239 @@ where
+    }
+
+    fn memfd_alloc(&self, mmap_size: usize) -> VhostUserResult<(*mut c_void, i64, i64)> {
+        let mut ret_val;
+        let inflight_file;
+        let dup_inflight_file;
+
+        // create an anonymous file and return a file descriptor to it.
+        // unsafe as extern syscall funciton is called
+        ret_val = unsafe {
+            libc::syscall(
+                libc::SYS_memfd_create,
+                &std::ffi::CString::new("inflight-region").unwrap(),
+                libc::MFD_ALLOW_SEALING,
+            )
+        };
+
+        if ret_val == -1 {
+            return Err(VhostUserError::MemFdCreateError);
+        }
+
+        inflight_file = ret_val;
+
+        // truncate the inflight_file to "mmap_size" length.
+        // unsafe as extern "ftruncate" is called
+        ret_val = unsafe { libc::ftruncate(inflight_file as RawFd, mmap_size as i64) } as i64;
+
+        if ret_val == -1 {
+            return Err(VhostUserError::FileTrucateError);
+        }
+
+        // place seals to restrict further modifications to the inflight_file file.
+        // unsafe as extern fcntl is called.
+        ret_val = unsafe {
+            libc::fcntl(
+                inflight_file as RawFd,
+                libc::F_ADD_SEALS,
+                libc::F_SEAL_GROW | libc::F_SEAL_SHRINK | libc::F_SEAL_SEAL,
+            )
+        } as i64;
+
+        if ret_val == -1 {
+            return Err(VhostUserError::MemFdSealError);
+        }
+
+        // duplicate the inflight_file file descriptor.
+        // unsafe as extern dup is called.
+        ret_val = unsafe { libc::dup(inflight_file as RawFd).into() };
+
+        if ret_val == -1 {
+            return Err(VhostUserError::MemFdCreateError);
+        }
+
+        dup_inflight_file = ret_val;
+
+        // map the inflight_file into memory.
+        // unsafe as extern mmap is called
+        Ok((
+            unsafe {
+                libc::mmap(
+                    std::ptr::null_mut(),
+                    mmap_size,
+                    libc::PROT_READ | libc::PROT_WRITE,
+                    libc::MAP_SHARED,
+                    inflight_file as RawFd,
+                    0,
+                )
+            },
+            inflight_file,
+            dup_inflight_file,
+        ))
+    }
+
+    fn set_desc_num_packed(&mut self, inflight_region: u64, num_queues: u16, queue_size: u16) {
+        let raw_ptr = inflight_region as *mut QueueRegionPacked;
+
+        for i in 0..num_queues {
+            unsafe {
+                // safe as we ensure that the raw_ptr is valid and that the correct amount of memory is
+                // mapped.
+                let queue_region = raw_ptr.offset(i as isize).as_mut().unwrap();
+                queue_region.desc_num = queue_size;
+            }
+        }
+    }
+
+    fn set_desc_num_split(&mut self, inflight_region: u64, num_queues: u16, queue_size: u16) {
+        let raw_ptr = inflight_region as *mut QueueRegionSplit;
```

Comment:
How about this for code readability:
```
let mut regions = unsafe { slice::from_raw_parts_mut(inflight_region as *mut QueueRegionSplit, num_queues as usize) };
regions.iter_mut().map(|v| v.desc_num = queue_size);
```

Reply:
Rust complains about unsused Map because of lazy iterators so I changed it slightly, let me know what you think.